pkg/gardenlet/controller/backupentry/backup_entry_migration_controller.go

```diff
@@ -0,0 +1,198 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func (c *Controller) backupEntryMigrationAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	c.backupEntryMigrationQueue.Add(key)
+}
+
+func (c *Controller) backupEntryMigrationUpdate(oldObj, newObj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		return
+	}
+
+	backupEntry, ok := newObj.(*gardencorev1beta1.BackupEntry)
+	if !ok {
+		return
+	}
+	if backupEntry.Generation == backupEntry.Status.ObservedGeneration && !v1beta1helper.HasOperationAnnotation(backupEntry.ObjectMeta) {
+		return
+	}
+
+	c.backupEntryMigrationQueue.Add(key)
+}
+
+func (c *Controller) backupEntryMigrationDelete(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	c.backupEntryMigrationQueue.Add(key)
+}
+
+// newMigrationReconciler returns an implementation of reconcile.Reconciler that forces the backup entry's restoration
+// to this seed during control plane migration if the preparation for migration in the source seed is not finished
+// after a certain grace period and is considered unlikely to succeed ("bad case" scenario).
+func newMigrationReconciler(
+	clientMap clientmap.ClientMap,
+	logger logrus.FieldLogger,
+	config *config.GardenletConfiguration,
+) reconcile.Reconciler {
+	return &migrationReconciler{
+		clientMap: clientMap,
+		logger:    logger,
+		config:    config,
+	}
+}
+
+type migrationReconciler struct {
+	clientMap clientmap.ClientMap
+	logger    logrus.FieldLogger
+	config    *config.GardenletConfiguration
+}
+
+func (r *migrationReconciler) Reconcile(ctx context.Context, req reconcile.Request) (result reconcile.Result, err error) {
+	log := r.logger.WithField("backupentry", req.String())
+
+	gardenClient, err := r.clientMap.GetClient(ctx, keys.ForGarden())
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get garden client: %w", err)
+	}
+
+	backupEntry := &gardencorev1beta1.BackupEntry{}
+	if err := gardenClient.Client().Get(ctx, req.NamespacedName, backupEntry); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Infof("[BACKUPENTRY MIGRATION] Skipping because BackupEntry has been deleted")
+			return reconcile.Result{}, nil
+		}
+		log.Infof("[BACKUPENTRY MIGRATION] Unable to retrieve object from store: %+v", err)
+		return reconcile.Result{}, err
+	}
+
+	// If the backup entry is being deleted or no longer being migrated to this seed, clear migration start time and don't requeue
+	if backupEntry.DeletionTimestamp != nil || !controllerutils.BackupEntryIsBeingMigratedToThisGardenlet(backupEntry, r.config) {
+		log.Debugf("[BACKUPENTRY MIGRATION] Clearing migration start time")
+		if err := setMigrationStartTime(ctx, gardenClient.Client(), backupEntry, nil); err != nil {
```

Comment:
Am I seeing this correctly, that this will cause an empty patch request for almost every backupentry reconciliations? (same applies to shoot counterpart, obviously)

Reply:
Why? From my perspective this code will be executed exactly once for each shoot - when the shoot is either deleted or no longer being migrated to the seed this gardenlet is responsible for (both seed names are equal, either the `migrate` phase finished successfully and `restore` took over, or this controller forced the restore). At this point, we'll clear the migration start time and forget about the shoot (we return without requeue and the informer will stop informing us about this shoot).

Reply:
Sorry, you are right. I missed the `BackupEntryMigrationFilterFunc` in the event handler...